Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift

```diff
@@ -1257,7 +1257,29 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
           node.closeQuote.id,
         ].compactMap { $0 }
       )
+    } else if (node.openQuote.tokenKind == .stringQuote && node.openQuote.presence == .missing),
+      (node.closeQuote.tokenKind == .stringQuote && node.closeQuote.presence == .missing),
+      !node.segments.isMissingAllTokens
```

Comment:
There’s no need to use `&&` or parentheses here. `,` in an `if` statement is a conjunction already. 

---

Also, I don’t think it’s necessary to check for the token kind of `openQuote` and `closeQuote`. The parser doesn’t generate any trees that hit this at the moment but in theory the diagnostic should also work for multi-line string literals where the `tokenKind` is `multilineStringQuote`. You’d just need to pass the kind of the quote to the error message.

Reply:
I just wanted to express the conditions for both the openQuote and closeQuote separately 😁
As there is no particular significance, I will make the change and differentiate them using a comma!

----

Hmm... If we don't check types of quotes here or check if a newline is included in the trivia of the quote token, this test case will fail.

```swift
  func testRawStringErrors6() {
    assertParse(
      ###"""
      let _ = ##"""1️⃣aa
        foobar
        aa2️⃣"""##
      """###,
      diagnostics: [
        DiagnosticSpec(locationMarker: "1️⃣", message: "multi-line string literal content must begin on a new line", fixIts: ["insert newline"]),
        DiagnosticSpec(locationMarker: "2️⃣", message: "multi-line string literal closing delimiter must begin on a new line", fixIts: ["insert newline"]),
      ],
      fixedSource: ###"""
        let _ = ##"""
          aa
          foobar
          aa
          """##
        """###
    )
  }
```

In the given case, both the openQuote and closeQuote tokens are marked as missing, but in reality, the quote exists with the missing only newline.

Are there cases in multiline string expressions where both quotes are missing and we need to emit a diagnostic similar to the one we just added, suggesting insert `"""` to surrounding them?

Or should we generate a single diagnostic here for the missing of a newline, instead of `handleInvalidMultilineStringQuote` called from `handleMissingToken`?

Reply:
Oh, I see. In that case maybe check that `unexpectedBeforeOpenQuote` and `unexpectedBeforeCloseQuote` are `nil` because they will contain the string quotes without a newline in the test case above.

Reply:
Ah, this works as intended and is more general than checking the token kind! Thank you for providing your feedback.